File: java/omnia/data/structure/mutable/HashMap.java
package omnia.data.structure.mutable;

import omnia.data.structure.Map;

public class HashMap<K, V> extends MaskingMap<K, V> {

  public static <K, V> HashMap<K, V> create() {
    return new HashMap<>();
  }

  public static <K, V> HashMap<K, V> copyOf(Map<? extends K, ? extends V> original) {
    return new HashMap<>(
        original.entries().stream()
            .collect(java.util.stream.Collectors.toMap(Map.Entry::key, Map.Entry::value)));
  }

  private HashMap() {
    super(new java.util.HashMap<>());
  }

  private HashMap(java.util.Map<? extends K, ? extends V> original) {
    super(new java.util.HashMap<>(original));
  }
}

File: src/omnia/data/structure/mutable/HashMap.kt
package omnia.data.structure.mutable

import omnia.data.structure.Map

class HashMap<K : Any, V : Any> private constructor() :
  MaskingMap<K, V>(kotlin.collections.HashMap()) {

  companion object {

    @JvmStatic
    fun <K : Any, V : Any> create(): HashMap<K, V> {
      return HashMap()
    }

    @JvmStatic
    fun <K : Any, V : Any> copyOf(original: Map<out K, out V>): HashMap<K, V> {
      return original.entries().toHashMap({ it.key() }, { it.value() })
    }

    @JvmStatic
    fun <K : Any, V : Any, E> Iterable<E>.toHashMap(
      keyMapper: (E) -> K,
      valueMapper: (E) -> V
    ): HashMap<K, V> {
      val map = create<K, V>()
      this.forEach { map.putMapping(keyMapper(it), valueMapper(it)) }
      return map
    }
  }
}